sdk/python/kfp/v2/dsl/artifact.py

```diff
@@ -0,0 +1,332 @@
+  _ALLOWED_MLMD_TYPES = {
+      PropertyType.INT: pipeline_spec_pb2.PrimitiveType.INT,
+      PropertyType.DOUBLE: pipeline_spec_pb2.PrimitiveType.DOUBLE,
+      PropertyType.STRING: pipeline_spec_pb2.PrimitiveType.STRING,
+  }
+
+  def __init__(self, type, description: str = None):
+    if type not in Property._ALLOWED_MLMD_TYPES:
+      raise ValueError('Property type must be one of %s.' %
+                       list(Property._ALLOWED_MLMD_TYPES.keys()))
+    self.type = type
+    self.description = description
+
+  @classmethod
+  def from_dict(cls, dict_data: Dict[str, Any]) -> 'Property':
+    """Deserializes the Property object from YAML dict."""
+    if not dict_data.get('type'):
+      raise TypeError('Missing type keyword in property dict.')
+    if dict_data['type'] == 'string':
+      kind = PropertyType.STRING
+    elif dict_data['type'] == 'int':
+      kind = PropertyType.INT
+    elif dict_data['type'] == 'double':
+      kind = PropertyType.DOUBLE
+    else:
+      raise TypeError('Got unknown type: %s' % dict_data['type'])
+
+    return Property(
+        type=kind,
+        description=dict_data['description']
+    )
+
+  def ir_type(self):
+    """Gets the IR primitive type."""
+    return Property._ALLOWED_MLMD_TYPES[self.type]
+
+  def get_type_name(self):
+    """Gets the type name used in YAML instance."""
+    if self.type == PropertyType.INT:
+      return 'int'
+    elif self.type == PropertyType.DOUBLE:
+      return 'double'
+    elif self.type == PropertyType.STRING:
+      return 'string'
+    else:
+      raise TypeError('Unexpected property type: %s' % self.type)
+
+
+class Artifact(object):
+  """KFP Artifact Python class.
+
+  Artifact Python class/object mainly serves following purposes in different
+  period of its lifecycle.
+
+  1. During compile time, users can use Artifact class to annotate I/O types of
+     their components.
+  2. At runtime, Artifact objects provide helper function/utilities to access
+     the underlying RuntimeArtifact pb message, and provide additional layers
+     of validation to ensure type compatibility.
+  """
+
+  # Name of the Artifact type.
+  TYPE_NAME = None
+  # Property schema.
+  # Example usage:
+  #
+  # PROPERTIES = {
+  #   'span': Property(type=PropertyType.INT),
+  #   # Comma separated of splits for an artifact. Empty string means artifact
+  #   # has no split.
+  #   'split_names': Property(type=PropertyType.STRING),
+  # }
+  PROPERTIES = None
+
+  # Initialization flag to support setattr / getattr behavior.
+  _initialized = False
+
+  def __init__(self, instance_schema: Optional[str] = None):
+    """Constructs an instance of Artifact"""
+    if self.__class__ == Artifact:
+      if not instance_schema:
+        raise ValueError(
+            'The "instance_schema" argument must be passed to specify a '
+            'type for this Artifact.')
+      schema = yaml.safe_load(instance_schema)['properties']
```

Comment:
What if `instance_schema` is invalid? Should we check for `properties in ...`?

Reply:
Good point. Added a check